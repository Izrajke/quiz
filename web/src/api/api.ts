import { toast } from 'react-toastify';

export const BACKEND_URL = 'http://localhost:8080';

/** Справочники. */
export enum DICTIONARIES {
  /** Фильтр по типу пака */
  filter = 'filter',
}

export interface ApiParams {
  /** Информация о запросе */
  input: RequestInfo;
  /** Дополнения к запросу */
  init?: RequestInit;
}

export interface SuccessResponse {
  success: boolean;
}

export const api = async <T = unknown>({
  input,
  init,
}: ApiParams): Promise<T | undefined> => {
  try {
    const response = await fetch(input, {
      ...init,
      headers: {
        ...init?.headers,
      },
    });

    if (response.ok) {
      return await response['json']();
    } else {
      toast.error(String(`${response.status} - ${response.statusText}`));
    }
  } catch (error) {
    toast.error('Проблема при подключении к серверу');
  }
};
