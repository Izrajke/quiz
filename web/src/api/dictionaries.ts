export interface Dictionary {
  uuid: string;
}

export interface DictionaryPackTypes extends Dictionary {
  id: number;
  title: string;
}
