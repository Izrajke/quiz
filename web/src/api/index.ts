export * from './socket';
export * from './quiz'
export * from './api';
