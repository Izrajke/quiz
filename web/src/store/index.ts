export * from './store';
export * from './Room';
export * from './Player';
export * from './App';
export * from './RootStore';
export * from './Dictionaries';
export * from './Home';
