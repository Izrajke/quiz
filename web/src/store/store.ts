import { useContext } from 'react';

import { MobXProviderContext } from 'mobx-react';

import { AppStore } from './App';
import { RootStore, root } from './RootStore';
import { PlayerStore } from './Player';
import { RoomStore } from './Room';
import { DictionariesStore } from './Dictionaries';
import { CreatePackStore } from './CreatePack';

interface Store {
  /** Root store */
  root: RootStore;
  /** Приложение */
  app: AppStore;
  /** Игрок */
  player: PlayerStore;
  /** Комната */
  room: RoomStore;
  /** Словари */
  dictionaries: DictionariesStore;
  /** Страница создания/просмотра пака */
  createPack: CreatePackStore;
}

export const store: Store = {
  root,
  app: root.app,
  player: root.player,
  room: root.room,
  dictionaries: root.dictionaries,
  createPack: root.createPack,
};

export const useStore = () => useContext(MobXProviderContext) as Store;
