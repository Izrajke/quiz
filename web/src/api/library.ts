import { api, BACKEND_URL } from 'api';

export interface NormalizedRangeQuestionsData {
  title: string;
  answer: number;
}

export interface NormalizedMultipleChoiceQuestionsData {
  title: string;
  options: string[];
  answer: number;
}

export interface NormalizedPackData {
  categoryId: number;
  title: string;
  pack: {
    rangeQuestions: NormalizedRangeQuestionsData[];
    multipleChoiceQuestions: NormalizedMultipleChoiceQuestionsData[];
  };
}

export const createPack = (pack: NormalizedPackData) => {
  const body = JSON.stringify(pack);

  return api({
    input: `${BACKEND_URL}/api/pack/create`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'POST',
      body,
    },
  });
};

export const loadPack = (id: string) => {
  return api({
    input: `${BACKEND_URL}/api/pack/view/${id}`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'POST',
    },
  });
};

export const editPack = (pack: NormalizedPackData, id: string) => {
  const body = JSON.stringify(pack);

  return api({
    input: `${BACKEND_URL}/api/pack/update/${id}`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'POST',
      body,
    },
  });
};

export const deletePack = (id: string) => {
  return api({
    input: `${BACKEND_URL}/api/pack/delete/${id}`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'POST',
    },
  });
};
