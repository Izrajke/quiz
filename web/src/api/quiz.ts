import { api, BACKEND_URL, DICTIONARIES } from 'api';

export interface CreateLobbyParams {
  name: string;
  players: number;
  password?: string;
}

export const createLobby = (params: CreateLobbyParams) => {
  const body = JSON.stringify(params);

  return api({
    input: `${BACKEND_URL}/create`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'POST',
      body,
    },
  });
};

export const loadDictionary = (dictionary: DICTIONARIES) => {
  return api({
    // TODO: подключить реальное api
    input: `/mock/${dictionary}.json`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'GET',
    },
  });
};

export interface LibraryItem {
  name: string;
  uuid: string;
  type: string;
  score: number;
}

export interface LibraryResponse {
  totalPages: number;
  content: LibraryItem[];
}

// TODO: подключить реальное api. POST запрос или GET через string params
export const loadLibrary = (page: number) => {
  // TODO: параметр страницы либо в Query либо переделать в POST
  console.log(page);
  return api<LibraryResponse>({
    input: `/mock/library.json`,
    init: {
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
      },
      method: 'GET',
    },
  });
};
