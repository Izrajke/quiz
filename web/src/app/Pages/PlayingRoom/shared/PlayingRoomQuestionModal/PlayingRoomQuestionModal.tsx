import { useCallback } from 'react';
import type { FunctionComponent } from 'react';

import { observer } from 'mobx-react-lite';

import { SocketResponseType } from 'api';

import { useStore } from 'store';

import { QuestionFirstType } from './QuestionFirstType';
import { QuestionSecondType } from './QuestionSecondType';

export const PlayingRoomQuestionModal: FunctionComponent = observer(() => {
  const { room } = useStore();

  const CurrentModal = useCallback(() => {
    switch (room.type) {
      case SocketResponseType.firstQuestionType:
        return <QuestionFirstType />;
      case SocketResponseType.secondQuestionType:
        return <QuestionSecondType />;
      default:
        return null;
    }
  }, [room.type]);

  return <CurrentModal />;
});

PlayingRoomQuestionModal.displayName = 'PlayingRoomQuestionModal';
