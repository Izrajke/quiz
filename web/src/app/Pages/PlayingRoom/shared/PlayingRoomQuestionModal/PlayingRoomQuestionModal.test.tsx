import { Provider } from 'mobx-react';
import { fireEvent, render, screen } from '@testing-library/react';

import { store } from 'store';

import { PlayingRoomQuestionModal } from './PlayingRoomQuestionModal';
import { SocketResponseType } from 'api';
import type { SocketAnswerData } from 'api';

describe('INTEGRATION: PlayingRoomQuestionModal', () => {
  const mockStore = { ...store };
  const { room } = mockStore;

  afterEach(() => {
    room.useQuestionModal(false);
  });

  beforeEach(() => {
    room.useQuestionModal(true);
  });

  const component = (
    <Provider {...mockStore}>
      <PlayingRoomQuestionModal />
    </Provider>
  );

  test('Компонент первого типа рендерится', () => {
    render(component);

    const question = {
      type: SocketResponseType.firstQuestionType,
      question: {
        title: 'Заголовок вопроса',
        options: {
          '1': 'Вариант 1',
          '2': 'Вариант 2',
          '3': 'Вариант 3',
          '4': 'Вариант 4',
        },
      },
    };

    room.useQuestionModal(true);
    room.setQuestion(question);
    const buttons = screen.getAllByRole('button');

    expect(screen.getByText(/заголовок вопроса/i)).toBeInTheDocument();
    buttons.forEach((button, index) => {
      expect(button.textContent).toBe(`Вариант ${index + 1}`);
    });
    expect(buttons).toMatchSnapshot();
  });

  test('Компонент второго типа рендерится', () => {
    render(component);

    const question = {
      type: SocketResponseType.secondQuestionType,
      question: {
        title: 'Заголовок вопроса',
      },
    };

    room.setQuestion(question);

    const header = screen.getByText(/заголовок вопроса/i);
    const input = screen.getByPlaceholderText(/ваш ответ/i);
    const button = screen.getByRole('button');

    expect(header).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    expect(header).toMatchSnapshot();
    expect(input).toMatchSnapshot();
    expect(button).toMatchSnapshot();
  });

  test('Компонент второго типа переходит в режим ожидания', () => {
    render(component);
    const input = screen.getByPlaceholderText(/ваш ответ/i);
    const button = screen.getByRole('button');

    fireEvent.change(input, { target: { value: '20' } });
    fireEvent.click(button);

    expect(
      screen.getByText(/Ожидание ответа остальных игроков.../i),
    ).toBeInTheDocument();
    expect(screen.getByText(/Ваш ответ:/i)).toBeInTheDocument();
    expect(screen.getByText(room.playerAnswer)).toBeInTheDocument();
    expect(room.playerAnswer).toBe('20');

    expect(screen.getByTestId('WaitingAnswerTextContainer')).toMatchSnapshot();
  });

  test('Компонент второго типа показывает результаты', () => {
    render(component);
    const answer: SocketAnswerData = {
      type: SocketResponseType.answerSecondQuestionType,
      answer: {
        value: '100',
      },
      options: [
        {
          color: 'player-2',
          name: 'gogsh',
          value: 21,
          time: 14,
        },
        {
          color: 'player-1',
          name: 'вася',
          value: 1,
          time: 5,
        },
      ],
    };

    room.setAnswer(answer);

    expect(room.answer).toBe(answer.answer.value);
    expect(screen.getByText('gogsh')).toBeInTheDocument();
    expect(screen.getByText('21')).toBeInTheDocument();
    expect(screen.getByText('14 c')).toBeInTheDocument();
    expect(screen.getByText(room.answer)).toBeInTheDocument();

    const answerBars = screen.getAllByTestId('AnswerBar');
    expect(answerBars.length).toBe(2);

    expect(answerBars).toMatchSnapshot();
  });
});
