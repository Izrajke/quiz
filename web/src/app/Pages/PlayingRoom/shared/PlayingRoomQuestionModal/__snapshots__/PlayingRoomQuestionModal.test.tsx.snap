// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент второго типа переходит в режим ожидания 1`] = `
<div
  class="answerContainer"
  data-testid="WaitingAnswerTextContainer"
>
  <div
    class="white-70 caption-1 weight-regular opacity-100 lh"
  >
    Ваш ответ:
  </div>
  <div
    class="white caption-1 weight-bold opacity-100 lh"
  >
    20
  </div>
</div>
`;

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент второго типа показывает результаты 1`] = `
Array [
  <div
    class="bar player-2"
    data-testid="AnswerBar"
    style="width: 21%; animation: AnswerBarWidth 3s ease;"
  >
    <div
      class="textContainer"
    >
      <div
        class="textContainer inside"
      >
        <div
          class="dark-2 text-0 weight-bold opacity-100 lh points"
        >
          21
        </div>
        <div
          class="dark-2 caption-2 weight-regular opacity-50 lh time"
        >
          14 c
        </div>
      </div>
    </div>
  </div>,
  <div
    class="bar player-1"
    data-testid="AnswerBar"
    style="width: 1%; animation: AnswerBarWidth 3s ease;"
  >
    <div
      class="textContainer"
    />
  </div>,
]
`;

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент второго типа рендерится 1`] = `
<div
  class="white text-2 weight-regular opacity-100 lh question"
>
  Заголовок вопроса
</div>
`;

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент второго типа рендерится 2`] = `
<input
  class="input"
  placeholder="Ваш ответ"
/>
`;

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент второго типа рендерится 3`] = `
<button
  class="button regular normal"
>
  Отправить
</button>
`;

exports[`INTEGRATION: PlayingRoomQuestionModal Компонент первого типа рендерится 1`] = `
Array [
  <button
    class="button regular normal button"
  >
    Вариант 1
  </button>,
  <button
    class="button regular normal button"
  >
    Вариант 2
  </button>,
  <button
    class="button regular normal button"
  >
    Вариант 3
  </button>,
  <button
    class="button regular normal button"
  >
    Вариант 4
  </button>,
]
`;
